File: src/main/scala/com/juanoff/scala/model/MultiList.scala
package com.juanoff.scala.model

//import com.juanoff.model.{DataStructure, DoWith, TestIt}
//import com.juanoff.types.Comparator

import com.juanoff.scala.types.Comparator

import scala.annotation.tailrec

class MultiList(val chunkSize: Int = 10) extends DataStructure {
  private var data: Vector[Vector[Any]] = Vector()
  private var totalSize: Int = 0

  override def add(value: Any): Unit = {
    if (data.isEmpty || data.last.size >= chunkSize) {
      data = data :+ Vector(value)
    } else {
      val lastChunk = data.last
      data = data.updated(data.size - 1, lastChunk :+ value)
    }
    totalSize += 1
  }

  override def get(index: Int): Any = {
    checkIndex(index)

    @tailrec
    def loop(chunks: Seq[Vector[Any]], currentIdx: Int): Any = {
      val head = chunks.head
      if (index < currentIdx + head.size) {
        head(index - currentIdx)
      } else {
        loop(chunks.tail, currentIdx + head.size)
      }
    }

    loop(data, 0)
  }

  override def insert(index: Int, value: Any): Unit = {
    checkIndexForInsert(index)

    if (index == totalSize) {
      add(value)
    } else {
      @tailrec
      def loop(i: Int, currentIdx: Int): Unit = {
        val chunk = data(i)
        if (index <= currentIdx + chunk.size) {
          val updatedChunk = chunk.patch(index - currentIdx, Seq(value), 0)
          data = data.updated(i, updatedChunk)
          totalSize += 1
          rebalanceIfNeeded()
        } else {
          loop(i + 1, currentIdx + chunk.size)
        }
      }

      loop(0, 0)
    }
  }

  override def remove(index: Int): Unit = {
    checkIndex(index)

    @tailrec
    def loop(i: Int, currentIdx: Int): Unit = {
      val chunk = data(i)
      if (index < currentIdx + chunk.size) {
        val updatedChunk = chunk.patch(index - currentIdx, Nil, 1)
        data = data.updated(i, updatedChunk)
        totalSize -= 1
        rebalanceIfNeeded()
      } else {
        loop(i + 1, currentIdx + chunk.size)
      }
    }

    loop(0, 0)
  }

  override def size(): Int = totalSize

  override def forEach(action: DoWith): Unit = data.flatten.foreach(item => action.doWith(item))

  override def firstThat(test: TestIt): Any = data.flatten.find(item => test.testIt(item)).orNull

  override def sort(comparator: Comparator): Unit = {
    val flatList: List[Any] = data.flatten.toList
    val sortedList = mergeSortFunctional(flatList, (o1: Any, o2: Any) => comparator.compare(o1, o2))
    rebuild(sortedList)
  }

  override def balance(): DataStructure = {
    val newList = new MultiList(chunkSize)
    data.flatten.foreach(newList.add)
    newList
  }

  override def clear(): Unit = while(size() > 0) remove(0)

  private def mergeSortFunctional(list: List[Any], comp: (Any, Any) => Int): List[Any] = {
    if (list.length <= 1) {
      list
    } else {
      val (left, right) = list.splitAt(list.length / 2)
      mergeLists(
        mergeSortFunctional(left, comp),
        mergeSortFunctional(right, comp),
        comp
      )
    }
  }

  private def mergeLists(left: List[Any], right: List[Any], comp: (Any, Any) => Int): List[Any] = {
    (left, right) match {
      case (Nil, _) => right
      case (_, Nil) => left
      case (lHead :: lTail, rHead :: rTail) =>
        if (comp(lHead, rHead) <= 0)
          lHead :: mergeLists(lTail, right, comp)
        else
          rHead :: mergeLists(left, rTail, comp)
    }
  }

  private def sortImperative(comp: (Any, Any) => Int): Unit = {
    val arr: Array[Any] = data.flatten.toArray
    val n = arr.length
    if (n <= 1) return

    val temp = new Array[Any](n)
    var width = 1

    while (width < n) {
      var i = 0
      while (i < n) {
        val left = i
        val mid = math.min(i + width, n)
        val right = math.min(i + 2 * width, n)

        var l = left
        var r = mid
        var k = left

        while (l < mid && r < right) {
          if (comp(arr(l), arr(r)) <= 0) {
            temp(k) = arr(l)
            l += 1
          } else {
            temp(k) = arr(r)
            r += 1
          }
          k += 1
        }

        System.arraycopy(arr, l, temp, k, mid - l)
        System.arraycopy(arr, r, temp, k + (mid - l), right - r)
        k = right

        var c = left
        while (c < right) {
          arr(c) = temp(c)
          c += 1
        }

        i += 2 * width
      }
      width *= 2
    }

    rebuild(arr.toSeq)
  }

  private def rebuild(flat: Seq[Any]): Unit = {
    data = Vector()
    totalSize = 0
    flat.foreach(add)
  }

  private def rebalanceIfNeeded(): Unit = {
    if (data.exists(_.isEmpty)) {
      rebuild(data.flatten.filter(_ != null))
    }
  }

  private def checkIndex(index: Int): Unit = {
    if (index < 0 || index >= totalSize) throw new IndexOutOfBoundsException(s"Index out of range: $index")
  }

  private def checkIndexForInsert(index: Int): Unit = {
    if (index < 0 || index > totalSize) throw new IndexOutOfBoundsException(s"Index out of range: $index")
  }
}

File: src/main/java/com/juanoff/model/MultiList.java
package com.juanoff.model;

import com.juanoff.types.Comparator;
import com.juanoff.util.Config;
import java.util.ArrayList;
import java.util.List;

public class MultiList implements DataStructure {
    private final int chunkSize;
    private final List<List<Object>> data = new ArrayList<>();

    private int totalSize = 0;

    public MultiList(int chunkSize) {
        if (chunkSize <= 0) {
            throw new IllegalArgumentException("Chunk size must be > 0");
        }
        this.chunkSize = chunkSize;
    }

    public MultiList() {
        this(Config.DEFAULT_MULTI_LIST_CHUNK_SIZE);
    }

    @Override
    public void add(Object value) {
        if (data.isEmpty() || data.getLast().size() >= chunkSize) {
            data.add(new ArrayList<>());
        }
        data.getLast().add(value);
        totalSize++;
    }

    @Override
    public Object get(int index) {
        checkIndex(index);

        int current = 0;
        for (List<Object> list : data) {
            if (index < current + list.size()) {
                return list.get(index - current);
            }
            current += list.size();
        }

        throw new IndexOutOfBoundsException();
    }

    @Override
    public void insert(int index, Object value) {
        checkIndexForInsert(index);

        if (index == totalSize) {
            add(value);
            return;
        }

        int current = 0;
        for (List<Object> list : data) {
            if (index <= current + list.size()) {
                list.add(index - current, value);
                totalSize++;
                rebalanceIfNeeded();
                return;
            }
            current += list.size();
        }

        add(value);
    }

    @Override
    public void remove(int index) {
        checkIndex(index);

        int current = 0;
        for (List<Object> list : data) {
            if (index < current + list.size()) {
                list.remove(index - current);
                totalSize--;
                rebalanceIfNeeded();
                return;
            }
            current += list.size();
        }
    }

    @Override
    public int size() {
        return totalSize;
    }

    @Override
    public void forEach(DoWith action) {
        data.forEach(list -> list.forEach(action::doWith));
    }

    @Override
    public Object firstThat(TestIt test) {
        for (List<Object> list : data) {
            for (Object obj : list) {
                if (test.testIt(obj)) {
                    return obj;
                }
            }
        }
        return null;
    }

    @Override
    public void sort(Comparator comparator) {
        List<Object> flat = flatten();
        mergeSort(flat, comparator);
        rebuild(flat);
    }

    @Override
    public DataStructure balance() {
        MultiList newList = new MultiList(chunkSize);
        List<Object> flat = flatten();
        for (Object o : flat) {
            newList.add(o);
        }
        return newList;
    }
    
    public void clear() {
        data.clear();
    }

    private List<Object> flatten() {
        List<Object> result = new ArrayList<>();
        data.forEach(result::addAll);
        return result;
    }

    private void rebuild(List<Object> flat) {
        data.clear();
        for (Object o : flat) {
            add(o);
        }
    }

    private void rebalanceIfNeeded() {
        if (data.size() < 2) {
            return;
        }

        if (data.stream().anyMatch(List::isEmpty)) {
            rebuild(flatten());
        }
    }

    private void mergeSort(List<Object> list, Comparator comp) {
        if (list.size() < 2) return;

        int mid = list.size() / 2;
        List<Object> left = new ArrayList<>(list.subList(0, mid));
        List<Object> right = new ArrayList<>(list.subList(mid, list.size()));

        mergeSort(left, comp);
        mergeSort(right, comp);

        merge(list, left, right, comp);
    }

    private void merge(List<Object> result, List<Object> left, List<Object> right, Comparator comp) {
        int i = 0, j = 0, k = 0;

        while (i < left.size() && j < right.size()) {
            if (comp.compare(left.get(i), right.get(j)) <= 0)
                result.set(k++, left.get(i++));
            else
                result.set(k++, right.get(j++));
        }

        while (i < left.size()) result.set(k++, left.get(i++));
        while (j < right.size()) result.set(k++, right.get(j++));
    }

    private void checkIndex(int index) {
        if (index < 0 || index >= size()) {
            throw new IndexOutOfBoundsException("Index out of range: " + index);
        }
    }

    private void checkIndexForInsert(int index) {
        if (index < 0 || index > size()) {
            throw new IndexOutOfBoundsException("Index out of range: " + index);
        }
    }
}
